### PR TITLE
GPU: CMake enable_language(HIP)

### PR DESCRIPTION
**Pull request by mconcas**
PR to track work state and discuss on the spot.

- Add `set_target_{CUDA,HIP}_arch()` function in FindO2GPU.cmake

Relevant changes so far for using HIP as a language:
- CMake version required >= 3.21
- Invoked compiler exe: `/opt/rocm/bin/hipcc` -> `/opt/rocm-X/llvm/bin/clang++`
- `*hip.cxx` -> `*.hip` renaming
- Explicit removal of autodetected integrated GPU architectures (e.g. `gfx1036`)

### CODE REVIEW

File: Detectors/ITSMFT/ITS/tracking/GPU/hip/.gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 # Don't track generated hip sources
-*.hip.cxx
\ No newline at end of file
+*.hip.cxx
```
Comment (davidrohr) [inline Detectors/ITSMFT/ITS/tracking/GPU/hip/.gitignore:2]: I think we should remove the *.hip.cxx then?

File: GPU/GPUTracking/Base/hip/CMakeLists.txt [hunk 2]
```diff
@@ -88,15 +88,4 @@ if(ALIGPU_BUILD_TYPE STREQUAL "Standalone")
...
-elseif(HIP_AMDGPUTARGET AND HIP_AMDGPUTARGET MATCHES "gfx908")
-  message(STATUS "Using optimized HIP settings for MI100 GPU")
-  target_compile_definitions(${targetName} PUBLIC GPUCA_GPUTYPE_MI2xx)
-elseif(HIP_AMDGPUTARGET AND HIP_AMDGPUTARGET MATCHES "gfx90a")
-  message(STATUS "Using optimized HIP settings for MI210 GPU")
-  target_compile_definitions(${targetName} PUBLIC GPUCA_GPUTYPE_MI2xx)
-else()
-  target_compile_definitions(${targetName} PUBLIC GPUCA_GPUTYPE_VEGA)
-endif()
+set_target_hip_arch(${targetName})
```
Comment (davidrohr) [inline GPU/GPUTracking/Base/hip/CMakeLists.txt:91]: please fix the tailing newlines at end of file

File: dependencies/FindO2GPU.cmake [hunk 3]
```diff
@@ -200,20 +230,27 @@ if(ENABLE_HIP)
...
+  if (CMAKE_HIP_COMPILER)
+    enable_language(HIP)
+    message(STATUS "HIP language enabled: ${CMAKE_HIP_COMPILER}")
+  endif()
   if(hip_FOUND AND hipcub_FOUND AND rocthrust_FOUND AND rocprim_FOUND AND hip_HIPCC_EXECUTABLE AND hip_HIPIFY_PERL_EXECUTABLE)
     set(HIP_ENABLED ON)
     set_target_properties(roc::rocthrust PROPERTIES IMPORTED_GLOBAL TRUE)
-    message(STATUS "HIP Found (${hip_HIPCC_EXECUTABLE} version ${hip_VERSION})")
+    message(STATUS "HIP Found (${hip_HIPCC_EXECUTABLE} version ${hip_VERSION})") # not needed anymore eventually if we migrate to native HIP clang++
     set(O2_HIP_CMAKE_CXX_FLAGS "-fgpu-defer-diag -mllvm -amdgpu-enable-lower-module-lds=false -Wno-invalid-command-line-argument -Wno-unused-command-line-argument -Wno-invalid-constexpr -Wno-ignored-optimization-argument -Wno-unused-private-field -Wno-pass-failed")
```
Comment (davidrohr) [inline dependencies/FindO2GPU.cmake:241]: Can't we put the O2_HIP_... flags now into CMAKE_HIP_FLAGS?
Though not so sure about the linker flags.

Comment (mconcas) [reply dependencies/FindO2GPU.cmake:241]: that's the plan, during the process I make them to coexist until everything is ported.

**Comment by mconcas** [comment]
@davidrohr: I added two new CMake macros to automatically create "hipified libraries/executables" out of cuda sources.
Essentially, they wrap all the boilerplate inside a function that eventually forwards all into the usual `o2_add_library/executable`.
This cleans up CMakeLists files and makes this feature globally available.

Concerning the language enabling I am experiencing some weird behaviours on a bunch of different setups.
I wanted to use the `o2-gpu-memory-benchmark-hip -p239` command to check everything compiled in the new way worked as expected.
In 2 out of 3 setups this doesn't work, and it does raise different errors (probably driven by different rocm versions).
I then created a separate standalone project [here](https://github.com/mconcas/GPUCompilationTests) to isolate and reproduce the issue, and (un)-fortunately it does not present the issue at all, it just works.
Just to summarise, if this rings you any bell:

1) First box: **Ubuntu 22.04/ROCm v5.6.1/Radeon Pro WX 9100**
  a) Old implementation **works**.
  b) Reproducer **works**.
  b) New version compiled within O2 does **crash at first kernel invocation**

2) Second box: **EPN node: ROCm v5.5.3/Mi50/100**
  a) Old implementation works.
  b) Reproducer works.
  b) New version compiled within O2 does **crash at first kernel invocation with a more verbose error**, still probably same issue

3) Third box: **Ubuntu 22.04/ROCm v5.6.1/AMD Radeon Pro WX3100**
  a) Old implementation works.
  b) Reproducer works.
  b) New version compiled within O2 **works**.

I am a bit puzzled, I checked the compile and link commands, linked libraries, versions, fat binaries field in `readelf`, ...

Any suggestion is most welcome.

**Comment by mconcas** [comment]
@davidrohr I'd start with merging this. The issue breaking OCL build was related to misuse of the new function in ITStrackingGPU, where it is actually not required.

Later I'd rebase latest hybrid ITSGPU developments on this to test it, then gradually move to the GPUFramework stuff.

**Comment by davidrohr** [comment]
ok, we can merge it. Do you want to rebase and squash the cleanup and fix commits, or shall I just squash the whole PR?

**Comment by mconcas** [comment]
> ok, we can merge it. Do you want to rebase and squash the cleanup and fix commits, or shall I just squash the whole PR?

I'd go for a squash, we save aspin of CI.

**Pull request merged**